Python/Product/Debugger/Debugger/DebugConnection.cs

```diff
@@ -93,7 +94,7 @@ class DebugConnection : IDisposable {
             _eventThread.Name = "Python Debugger Event Handling " + _processGuid;
             _eventThread.Start();
 
-            _debuggerThread = new Thread(MessageProcessingThread);
+            _debuggerThread = new AsyncThread(MessageProcessingThreadAsync);
```

Comment:
I'm not sure I understand why do you need the thread at all. Before the change, code execution would've jumped into the thread pool and never returned back. `_debuggerThread.Join` was pretty much similar to `Task.Wait`. Did we have any bugs because of that?

Reply:
That's what it was doing, and it seemed to be deadlocking in the test (which is also jumping into the thread pool).

Having a dedicated thread to come back to after each real await is the point of the thread. This makes sure it actually works.

Reply:
Jumping into thread pool can't cause a deadlock by itself, while binding task continuations to a single thread does it with guarantee for any task that is synchronously waited on that same thread. Conversion from free-threaded to a single-threaded is a major breaking change.

What test is got blocked and what are the deadlock callstacks?

Reply:
So it's the debugger message processing thread that gets blocked, and I'm pretty sure since we're using slim concurrency objects instead of kernel level ones that the thread pool can try and schedule a task on an existing thread without realizing that it's blocked. 

Ensuring that debugger messages are all processed on the same thread doesn't generally change concurrency - it actually fixes the implementation to match the design, ensuring that messages _always_ have a thread to come back to that is only ever blocked by another message or by forcing a task to the UI thread. It's a single long-running task, not a transient free-threaded threadpool type task, and keeping it on its own thread is meant to isolate it from whatever other threadpool threads may be doing, thereby simplifying concurrency rather than complicating it.

Reply:
The tests are the AttachNewThread tests that wait for a breakpoint to be hit. While they're waiting for the breakpoint, no messages are being processed on the debugger thread (without this fix). Then when it stops waiting, all the messages are processed at once. Adding the timestamps and setting breakpoints in the message handling loop (and seeing which thread it kept moving to - not the one it's supposed to stay on!) is what told me the problem.

Reply:
Let me dig into that a bit. Is it only reproducible on VS16?

Reply:
I see it on VS15 as well. Make sure you add the timestamps to the log messages (as in this PR).